### PR TITLE
Fix or postpone remaining TODO(QUIC) items

### PR DESCRIPTION
**Pull request by t8m**
Postpone two TODO(QUIC) items appropriately
 
- The one in ch_rx_handle_packet() is a tuning thing -> QUIC FUTURE
- The one in ossl_quic_tserver_shutdown() is a server thing -> QUIC SERVER

Add handling of SSL_get_shutdown()
    
- Return SSL_SENT_SHUTDOWN and SSL_RECEIVED_SHUTDOWN with semantics
  similar to TLS connections.

### CODE REVIEW

File: doc/man3/SSL_set_shutdown.pod [hunk 1]
```diff
@@ -57,13 +57,20 @@ If a close_notify was received, SSL_RECEIVED_SHUTDOWN will be set,
...
 
 =head1 RETURN VALUES
 
 SSL_set_shutdown() does not return diagnostic information.
 
-SSL_get_shutdown() returns the current setting.
+SSL_get_shutdown() returns the current shutdown state as set or based
+on the actual connection state.
+
+SSL_get_shutdown() returns 0 if called on a QUIC stream SSL object. It returns
```
Comment (mattcaswell) [inline doc/man3/SSL_set_shutdown.pod:69]: Perhaps for the second sentence we should start with "If it is called on a QUIC connection SSL object, then it returns a value with..."

File: test/quicapitest.c [hunk 2]
```diff
@@ -1335,6 +1335,50 @@ static int test_alpn(int idx)
...
+
+    if (!TEST_int_eq(SSL_shutdown(clientquic), 0))
+        goto err;
+
+    if (!TEST_int_eq(SSL_get_shutdown(clientquic), SSL_SENT_SHUTDOWN))
+        goto err;
+
+    do {
+        ossl_quic_tserver_tick(qtserv);
+        OSSL_sleep(1);
```
Comment (mattcaswell) [inline test/quicapitest.c:1366]: Why is this necessary?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/quicapitest.c (hunk 2 above)
Comment (t8m) [reply test/quicapitest.c:1366]: I just did not want it to be a full busy wait when it is waiting for the draining timeout to expire.

Comment (mattcaswell) [reply test/quicapitest.c:1366]: Ah...right. In that case I'd suggest using the fake timer for this, so we don't have to pause in the test waiting for this.

File: doc/man3/SSL_set_shutdown.pod (hunk 1 above)
Comment (t8m) [reply doc/man3/SSL_set_shutdown.pod]: Done.

**Comment by t8m** [comment]
@mattcaswell please look again

**Comment by mattcaswell** [comment]
> @mattcaswell please look again

What about using the fake timer in the test?

File: ssl/quic/quic_impl.c [hunk 3]
```diff
@@ -3571,6 +3571,27 @@ const SSL_CIPHER *ossl_quic_get_cipher(unsigned int u)
...
+int ossl_quic_get_shutdown(const SSL *s)
+{
+    QCTX ctx;
+    int shut = 0;
+
+    if (!expect_quic_conn_only(s, &ctx))
+        return 0;
+
+    if (ossl_quic_channel_is_term_any(ctx.qc->ch)) {
+        shut |= SSL_SENT_SHUTDOWN;
```
Comment (hlandau) [inline ssl/quic/quic_impl.c:3587]: Is this really correct? It seems like we should vary it based on whether the TCAUSE was local or remote. Depending on that these flags should appear in the opposite chronological order based on these two predicates.

**Review by hlandau — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_impl.c (hunk 3 above)
Comment (t8m) [reply ssl/quic/quic_impl.c:3587]: IMO it is correct. When we receive the CONNECTION_CLOSE frame we immediately reply with connection close too and enter the DRAINING state. I do not see how to implement it otherwise.

**Comment by t8m** [comment]
@mattcaswell I've changed the test to use fake time, please reconfirm

**Comment by mattcaswell** [comment]
Hmmm. CI failure is disturbing (but probably not related)

**Comment by t8m** [comment]
Yeah, definitely not related to this as this does not touch the tserver test and the added api is also not invoked by that test.

**Comment by t8m** [comment]
Added #22424 issue for the intermittent CI failure.

**Review by hlandau — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed to master.
